File: todo-cli-java/src/main/java/com/todo/Storage.java
package com.todo;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

/**
 * Handles storage operations for tasks using JSON file.
 */
public class Storage {
    private final String filePath;
    private final Gson gson;
    
    /**
     * Creates a new Storage instance.
     * 
     * @param filePath Path to the JSON file
     */
    public Storage(String filePath) throws IOException {
        this.filePath = filePath;
        this.gson = new GsonBuilder().setPrettyPrinting().create();
        
        // Create file if it doesn't exist
        Path path = Paths.get(filePath);
        if (!Files.exists(path)) {
            saveTasks(new ArrayList<>());
        }
    }
    
    /**
     * Adds a new task to storage.
     */
    public void addTask(Task task) throws IOException {
        List<Task> tasks = loadTasks();
        tasks.add(task);
        saveTasks(tasks);
    }
    
    /**
     * Gets all tasks from storage.
     */
    public List<Task> getAllTasks() throws IOException {
        return loadTasks();
    }
    
    /**
     * Marks a task as completed by its ID.
     */
    public void completeTask(String taskId) throws IOException, TaskNotFoundException {
        List<Task> tasks = loadTasks();
        boolean found = false;
        
        for (Task task : tasks) {
            if (task.getId().equals(taskId)) {
                task.markCompleted();
                found = true;
                break;
            }
        }
        
        if (!found) {
            throw new TaskNotFoundException("Task with ID " + taskId + " not found");
        }
        
        saveTasks(tasks);
    }
    
    /**
     * Deletes a task by its ID.
     */
    public void deleteTask(String taskId) throws IOException, TaskNotFoundException {
        List<Task> tasks = loadTasks();
        int initialSize = tasks.size();
        
        tasks.removeIf(task -> task.getId().equals(taskId));
        
        if (tasks.size() == initialSize) {
            throw new TaskNotFoundException("Task with ID " + taskId + " not found");
        }
        
        saveTasks(tasks);
    }
    
    /**
     * Gets a task by its ID.
     */
    public Task getTaskById(String taskId) throws IOException, TaskNotFoundException {
        List<Task> tasks = loadTasks();
        
        for (Task task : tasks) {
            if (task.getId().equals(taskId)) {
                return task;
            }
        }
        
        throw new TaskNotFoundException("Task with ID " + taskId + " not found");
    }
    
    /**
     * Loads tasks from the JSON file.
     */
    private List<Task> loadTasks() throws IOException {
        try (FileReader reader = new FileReader(filePath)) {
            Type taskListType = new TypeToken<ArrayList<Task>>() {}.getType();
            List<Task> tasks = gson.fromJson(reader, taskListType);
            return tasks != null ? tasks : new ArrayList<>();
        }
    }
    
    /**
     * Saves tasks to the JSON file.
     */
    private void saveTasks(List<Task> tasks) throws IOException {
        try (FileWriter writer = new FileWriter(filePath)) {
            gson.toJson(tasks, writer);
        }
    }
    
    /**
     * Exception thrown when a task is not found.
     */
    public static class TaskNotFoundException extends Exception {
        public TaskNotFoundException(String message) {
            super(message);
        }
    }
}

File: todo-cli/src/storage.rs
use crate::task::Task;
use anyhow::Result;
use serde_json;
use std::fs::File;
use std::io::{BufReader, BufWriter};
use std::path::Path;

pub struct Storage {
    file_path: String,
}

impl Storage {
    pub fn new(file_path: &str) -> Result<Self> {
        let storage = Storage {
            file_path: file_path.to_string(),
        };
        
        // Create file if it doesn't exist
        if !Path::new(file_path).exists() {
            let empty_tasks: Vec<Task> = Vec::new();
            storage.save_tasks(&empty_tasks)?;
        }
        
        Ok(storage)
    }

    pub fn add_task(&mut self, task: Task) -> Result<()> {
        let mut tasks = self.load_tasks()?;
        tasks.push(task);
        self.save_tasks(&tasks)
    }

    pub fn get_all_tasks(&self) -> Result<Vec<Task>> {
        self.load_tasks()
    }

    pub fn complete_task(&mut self, task_id: &str) -> Result<()> {
        let mut tasks = self.load_tasks()?;
        
        for task in tasks.iter_mut() {
            if task.id == task_id {
                task.mark_completed();
                return self.save_tasks(&tasks);
            }
        }
        
        Err(anyhow::anyhow!("Task with ID {} not found", task_id))
    }

    pub fn delete_task(&mut self, task_id: &str) -> Result<()> {
        let mut tasks = self.load_tasks()?;
        let initial_len = tasks.len();
        
        tasks.retain(|task| task.id != task_id);
        
        if tasks.len() == initial_len {
            return Err(anyhow::anyhow!("Task with ID {} not found", task_id));
        }
        
        self.save_tasks(&tasks)
    }

    pub fn get_task_by_id(&self, task_id: &str) -> Result<Task> {
        let tasks = self.load_tasks()?;
        
        for task in tasks {
            if task.id == task_id {
                return Ok(task);
            }
        }
        
        Err(anyhow::anyhow!("Task with ID {} not found", task_id))
    }

    pub fn update_task(&mut self, updated_task: Task) -> Result<()> {
        let mut tasks = self.load_tasks()?;
        
        for task in tasks.iter_mut() {
            if task.id == updated_task.id {
                *task = updated_task;
                return self.save_tasks(&tasks);
            }
        }
        
        Err(anyhow::anyhow!("Task with ID {} not found", updated_task.id))
    }

    fn load_tasks(&self) -> Result<Vec<Task>> {
        let file = File::open(&self.file_path)?;
        let reader = BufReader::new(file);
        let tasks: Vec<Task> = serde_json::from_reader(reader)?;
        Ok(tasks)
    }

    fn save_tasks(&self, tasks: &Vec<Task>) -> Result<()> {
        let file = File::create(&self.file_path)?;
        let writer = BufWriter::new(file);
        serde_json::to_writer_pretty(writer, tasks)?;
        Ok(())
    }
}